### DISCUSSION TITLE
Simple TE3 Version Control for Workspace Model

### DISCUSSION THREAD

**Discussion by 10xape** [TabularEditor/TabularEditor3#1421] [Data Modeling]
Hi

I want to establish simple (data pipelin free) foir incremental refresh workspace mode model.

I was wondering if you could give me a same idea on the best simple approach and it does not have to be perfect as well follow the best practices, but have to work in case I want to revert to a previous version without reauploding entire desktop model back to the Power BI service (very time consuming) 

I do not miond to do it in a very manual way such as open workspace mode and save it as new version before making any new changes.

Would that work with coming back to the previous TE3 saved file to retrive  previous metadata state of the model ?

Thanks 

Matt

**Comment by otykier** [comment]
This is a difficult question to answer in a short post. It all depends on the scenarios you need to support (i.e. single-developer vs. multi-developer, model-only development vs. full-stack - i.e. ETL, semantic models and reports), deployment strategy, etc. But in general, doing version control comes down to choosing a version control system. These days, [git](https://git-scm.com/) is the only version control system worth investing in, as everybody uses it.

If you're not already familiar with git, [here is a good place to start](https://code.visualstudio.com/docs/sourcecontrol/intro-to-git) (you don't have to use VS Code, but it provides a great UI for git operations).

Once you're familiar with git, there are multiple approaches depending on whether or not you're using Microsoft Fabric and planning to use the git integration supported by Fabric workspaces.

In any case, a lot of this goes a bit beyond Tabular Editor, as our tool is really just an editor for semantic model metadata - how you choose to version control it is up to you. However, we did write a guide specifically for [Fabric git integration](https://blog.tabulareditor.com/2024/04/11/tabular-editor-and-fabric-git-integration/), which may be helpful to you.
